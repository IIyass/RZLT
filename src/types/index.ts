import { ReactNode } from 'react';

export enum SnackBarType {
  ERROR = 'ERROR',
  SUCCESS = 'SUCCESS',
}

export interface GithubUser {
  name?: string;
  avatarUrl: string;
  bio?: string;
  email?: string;
  followers: number;
  following: number;
  publicRepos: number;
  login: string;
  htmlUrl: string;
}

export interface GithubUserRepo {
  name: string;
  htmlUrl: string;
  forksCount: number;
  createdAt: string;
  svnUrl: string;
}

export interface SearchHistory {
  name: string;
}

export interface RepoColumn {
  id: 'name' | 'createdAt' | 'htmlUrl' | 'forksCount' | 'svnUrl';
  label: string;
  minWidth?: number;
  convert?: (value: string) => string;
  icon?: () => ReactNode;
}

export interface HistoryColumn {
  id: 'name';
  label: string;
  minWidth?: number;
}

export enum SearchStep {
  USER = 'USER',
  REPOSITORIES = 'REPOSITORIES',
}
