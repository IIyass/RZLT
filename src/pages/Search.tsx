import { Button, CircularProgress, Container } from '@mui/material';
import SearchForm from 'modules/SearchForm';
import { Box } from '@mui/material';
import { createStyles, makeStyles } from '@mui/styles';
import { useDispatch, useSelector } from 'react-redux';
import {
  getGithubUser,
  getGithubUserRepos,
  selectGithubUser,
  selectGithubUserRepos,
  selectLoading,
  selectSearchTerm,
  selectSearchStep,
  setSearchStep,
} from 'state/user/slice';
import UserCard from 'components/Card';
import { useEffect, useState } from 'react';
import { repoColumns } from 'constants/columns';
import Table from 'modules/RepositoriesTable';
import { SearchStep } from 'types';

const useStyles = makeStyles(() =>
  createStyles({
    box: {
      width: '100%',
      margin: '0 auto',
      padding: '24px',

      '& .MuiPaper-root': {
        margin: 'auto',
      },
      '& button': {
        marginBottom: '20px',
      },
    },
  })
);

const Search = () => {
  const githubUser = useSelector(selectGithubUser);
  const githubUserRepos = useSelector(selectGithubUserRepos);
  const loading = useSelector(selectLoading);
  const searchStep = useSelector(selectSearchStep);
  const [page, setPage] = useState(1);
  const [rowsPerPage, setRowsPerPage] = useState(5);
  const classes = useStyles();
  const dispatch = useDispatch();
  const searchTerm = useSelector(selectSearchTerm);

  useEffect(() => {
    if (searchStep === SearchStep.REPOSITORIES && githubUser) {
      dispatch(
        getGithubUserRepos({ search: githubUser.login, rowsPerPage, page })
      );
    }
  }, [dispatch, rowsPerPage, page, searchStep, githubUser]);

  useEffect(() => {
    if (searchTerm) {
      dispatch(getGithubUser(searchTerm));
    }
  }, [dispatch, searchTerm]);

  const renderSearch = () => {
    if (!githubUser) return;
    switch (searchStep) {
      case SearchStep.USER:
        return (
          <UserCard
            name={githubUser.name}
            avatarUrl={githubUser.avatarUrl}
            bio={githubUser.bio}
            email={githubUser.email}
            followers={githubUser.followers}
            following={githubUser.following}
            login={githubUser.login}
            url={githubUser.htmlUrl}
          />
        );
      case SearchStep.REPOSITORIES:
        return (
          <Table
            data={githubUserRepos}
            page={page}
            rowsPerPage={rowsPerPage}
            count={githubUser.publicRepos}
            setPage={(page: number) => setPage(page)}
            setRowsPerPage={(row: number) => setRowsPerPage(row)}
            columns={repoColumns}
          />
        );
    }
  };

  return (
    <Container>
      <Box className={classes.box}>
        {searchStep === SearchStep.REPOSITORIES && (
          <Button
            variant="contained"
            color="secondary"
            onClick={() => dispatch(setSearchStep(SearchStep.USER))}
          >
            Go Back
          </Button>
        )}

        <SearchForm search={searchTerm} />

        {loading ? (
          <CircularProgress data-testid="loading_circle" />
        ) : (
          renderSearch()
        )}
      </Box>
    </Container>
  );
};

export default Search;
