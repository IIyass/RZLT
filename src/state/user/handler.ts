import { PayloadAction } from '@reduxjs/toolkit';
import { fetchGithubUser, fetchGithubUserRepos } from 'api';
import { call, put, takeLatest } from 'redux-saga/effects';
import { StorageKeys } from 'storage/StorageKeys';
import { GithubUser, SearchStep, SnackBarType } from 'types';
import {
  setLoading,
  setSnackBar,
  getGithubUser,
  setGithubUser,
  getGithubUserRepos,
  setGithubUserRepos,
  setSearchHistory,
  loadSearchHistory,
  setSearchStep,
} from './slice';

export function* getGithubUserSaga(action: PayloadAction<string>) {
  try {
    yield put(setSearchStep(SearchStep.USER));
    yield put(setLoading(true));
    const response: GithubUser = yield call(fetchGithubUser, action.payload);
    yield put(setGithubUser(response));
    yield put(setLoading(false));
    const searchHistory = window.localStorage.getItem(
      StorageKeys.SearchHistory
    );

    window.localStorage.setItem(
      StorageKeys.SearchHistory,
      JSON.stringify([
        ...JSON.parse(searchHistory ? `${searchHistory}` : '[]'),
        { name: action.payload },
      ])
    );
  } catch (error: any) {
    yield put(setLoading(false));
    yield put(
      setSnackBar({
        snackBarStatus: SnackBarType.ERROR,
        openSnackBar: true,
        messages: error.response.data.message,
      })
    );
  }
}

export function* getGithubUserRepoSaga(
  action: PayloadAction<{ rowsPerPage: number; page: number; search: string }>
) {
  try {
    yield put(setLoading(true));
    const { rowsPerPage, page, search } = action.payload;
    const response: unknown = yield call(
      fetchGithubUserRepos,
      search,
      rowsPerPage,
      page
    );
    yield put(setGithubUserRepos(response));
    yield put(setLoading(false));
  } catch (error) {
    yield put(setLoading(false));
    yield put(
      setSnackBar({
        snackBarStatus: SnackBarType.ERROR,
        openSnackBar: true,
        messages: 'Error accured, please try again',
      })
    );
  }
}

export function* loadSearchHistorySaga() {
  try {
    yield put(setLoading(true));

    const searchHistory = window.localStorage.getItem(
      StorageKeys.SearchHistory
    );
    yield put(setSearchHistory(JSON.parse(searchHistory ?? '[]').reverse()));
    yield put(setLoading(false));
  } catch (error) {
    yield put(setLoading(false));
    yield put(
      setSnackBar({
        snackBarStatus: SnackBarType.ERROR,
        openSnackBar: true,
        messages: 'Error accured, please try again',
      })
    );
  }
}

export function* watcherUserSagas() {
  yield takeLatest(getGithubUser.type, getGithubUserSaga);
  yield takeLatest(getGithubUserRepos.type, getGithubUserRepoSaga);
  yield takeLatest(loadSearchHistory.type, loadSearchHistorySaga);
}
