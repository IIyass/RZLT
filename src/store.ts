import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import createSagaMiddleware from 'redux-saga';
import { rootSaga } from 'rootSaga';
import user from 'state/user/slice';

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: {
    user: user,
  },

  middleware: [...getDefaultMiddleware({ thunk: false }), sagaMiddleware],
});
sagaMiddleware.run(rootSaga);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
