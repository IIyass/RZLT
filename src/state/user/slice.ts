import { createAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from 'store';
import {
  GithubUser,
  GithubUserRepo,
  SearchHistory,
  SearchStep,
  SnackBarType,
} from 'types';

interface SnackBar {
  snackBarStatus: SnackBarType | undefined;
  openSnackBar: boolean;
  messages?: any;
}

export interface UserState {
  loading: boolean;
  githubUser?: GithubUser;
  githubUserRepos: GithubUserRepo[];
  snackBar: SnackBar;
  step: SearchStep;
  searchHistory: SearchHistory[];
  searchTerm: string;
}

export const initialState: UserState = {
  loading: false,
  githubUser: undefined,
  githubUserRepos: [],
  snackBar: {
    openSnackBar: false,
    snackBarStatus: undefined,
    messages: null,
  },
  step: SearchStep.USER,
  searchHistory: [],
  searchTerm: '',
};
export const getGithubUser = createAction<string>('getGithubUser');

export const loadSearchHistory = createAction('loadSearchHistory');

export const getGithubUserRepos = createAction<{
  search: string;
  rowsPerPage: number;
  page: number;
}>('getGithubUserRepos');

export const user = createSlice({
  name: 'user',
  initialState,
  reducers: {
    setLoading: (state: UserState, action: PayloadAction<boolean>) => {
      state.loading = action.payload;
    },

    setSnackBar: (state: UserState, action: PayloadAction<SnackBar>) => {
      state.snackBar.snackBarStatus = action.payload.snackBarStatus;
      state.snackBar.openSnackBar = action.payload.openSnackBar;
      state.snackBar.messages = action.payload.messages;
    },

    setGithubUser: (state: UserState, action: PayloadAction<any>) => {
      state.githubUser = action.payload;
    },

    setGithubUserRepos: (state: UserState, action: PayloadAction<any>) => {
      state.githubUserRepos = action.payload;
    },
    setSearchStep: (state: UserState, action: PayloadAction<SearchStep>) => {
      state.step = action.payload;
    },

    setSearchHistory: (
      state: UserState,
      action: PayloadAction<SearchHistory[]>
    ) => {
      state.searchHistory = action.payload;
    },

    setSearchTerm: (state: UserState, action: PayloadAction<string>) => {
      state.searchTerm = action.payload;
    },
  },
});
export const {
  setLoading,
  setSnackBar,
  setGithubUser,
  setGithubUserRepos,
  setSearchStep,
  setSearchHistory,
  setSearchTerm,
} = user.actions;

export const selectSnackBar = (state: RootState) => ({
  status: state.user.snackBar.snackBarStatus,
  open: state.user.snackBar.openSnackBar,
  messages: state.user.snackBar.messages,
});
export const selectLoading = (state: RootState) => state.user.loading;
export const selectGithubUser = (state: RootState) => state.user.githubUser;
export const selectGithubUserRepos = (state: RootState) =>
  state.user.githubUserRepos;
export const selectSearchStep = (state: RootState) => state.user.step;
export const selectSearchHistory = (state: RootState) =>
  state.user.searchHistory;

export const selectSearchTerm = (state: RootState) => state.user.searchTerm;

export default user.reducer;
