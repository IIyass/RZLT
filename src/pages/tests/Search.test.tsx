import { render, screen } from '@testing-library/react';
import {
  createTheme,
  StyledEngineProvider,
  ThemeProvider,
} from '@mui/material';
import { RootState } from 'store';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { SearchStep, SnackBarType } from 'types';
import Search from 'pages/Search';
import { getGithubUser, getGithubUserRepos } from 'state/user/slice';

const mockedUsedNavigate = jest.fn();
const mockedUseDispatchFn = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockedUsedNavigate,
}));

jest.mock('react-redux', () => ({
  ...(jest.requireActual('react-redux') as any),
  useDispatch: () => mockedUseDispatchFn,
}));

const initialState: Pick<RootState, 'user'> = {
  user: {
    loading: false,
    githubUser: {
      name: '',
      avatarUrl: '',
      followers: 0,
      following: 0,
      publicRepos: 0,
      login: '',
      htmlUrl: '',
    },
    githubUserRepos: [],
    snackBar: {
      openSnackBar: true,
      snackBarStatus: SnackBarType.SUCCESS,
      messages: 'Success',
    },
    step: SearchStep.USER,
    searchHistory: [],
    searchTerm: '',
  },
};
const mockStore = configureStore();
const mockedStore = (state: RootState) => mockStore(state);

const renderSearchForm = (state: RootState) => {
  return render(
    <StyledEngineProvider injectFirst>
      <ThemeProvider theme={createTheme()}>
        <Provider store={mockedStore(state)}>
          <Search />
        </Provider>
      </ThemeProvider>
    </StyledEngineProvider>
  );
};

describe('Search Form', () => {
  it('should match snapshot', () => {
    const { container } = renderSearchForm({
      user: {
        ...initialState.user,
        githubUser: {
          name: 'ilyass',
          avatarUrl: 'avatar.svg',
          followers: 10,
          following: 20,
          publicRepos: 20,
          login: 'iiyass',
          htmlUrl: 'url',
        },
      },
    });
    expect(container).toMatchSnapshot();
  });

  it('should get user information, if search term is available initially', async () => {
    renderSearchForm({
      user: {
        ...initialState.user,
        searchTerm: 'ilyass',
      },
    });
    expect(mockedUseDispatchFn).toHaveBeenCalledWith(getGithubUser('ilyass'));
  });

  it('should get user repositories information', async () => {
    renderSearchForm({
      user: {
        ...initialState.user,
        step: SearchStep.REPOSITORIES,
        githubUser: {
          name: 'ilyass',
          avatarUrl: 'avatar.svg',
          followers: 10,
          following: 20,
          publicRepos: 20,
          login: 'iiyass',
          htmlUrl: 'url',
        },
      },
    });
    expect(mockedUseDispatchFn).toHaveBeenCalledWith(
      getGithubUserRepos({ search: 'iiyass', rowsPerPage: 5, page: 0 })
    );
  });

  it('should render Go Back button, if search is repositories', () => {
    renderSearchForm({
      user: {
        ...initialState.user,
        step: SearchStep.REPOSITORIES,
        githubUser: {
          name: 'ilyass',
          avatarUrl: 'avatar.svg',
          followers: 10,
          following: 20,
          publicRepos: 20,
          login: 'iiyass',
          htmlUrl: 'url',
        },
      },
    });
    expect(screen.getByText('Go Back')).toBeInTheDocument();
  });

  it('should not render Go Back button, if search is user data', () => {
    renderSearchForm({
      user: {
        ...initialState.user,
        step: SearchStep.USER,
        githubUser: {
          name: 'ilyass',
          avatarUrl: 'avatar.svg',
          followers: 10,
          following: 20,
          publicRepos: 20,
          login: 'iiyass',
          htmlUrl: 'url',
        },
      },
    });
    expect(screen.queryByText('Go Back')).not.toBeInTheDocument();
  });

  it('should render loading circle, if data is not available yet.', async () => {
    renderSearchForm({
      user: {
        ...initialState.user,
        loading: true,
      },
    });
    expect(screen.getByTestId('loading_circle')).toBeInTheDocument();
  });

  it('should render user card, if seach step is USER', async () => {
    renderSearchForm({
      user: {
        ...initialState.user,
        step: SearchStep.USER,
        githubUser: {
          name: 'ilyass',
          avatarUrl: 'avatar.svg',
          followers: 10,
          following: 20,
          publicRepos: 20,
          login: 'iiyass',
          htmlUrl: 'url',
        },
      },
    });
    expect(screen.getByTestId('user_card')).toBeInTheDocument();
  });

  it('should render repositories table, if seach step is REPOSITORIES', async () => {
    renderSearchForm({
      user: {
        ...initialState.user,
        step: SearchStep.REPOSITORIES,
        githubUser: {
          name: 'ilyass',
          avatarUrl: 'avatar.svg',
          followers: 10,
          following: 20,
          publicRepos: 20,
          login: 'iiyass',
          htmlUrl: 'url',
        },
      },
    });
    expect(screen.getByTestId('repos_table')).toBeInTheDocument();
  });
});
