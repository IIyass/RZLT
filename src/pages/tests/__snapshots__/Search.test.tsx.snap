// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Form should match snapshot 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <div
      class="makeStyles-box-1 MuiBox-root css-0"
    >
      <div
        class="makeStyles-box-2 MuiBox-root css-0"
      >
        <form
          action="#"
        >
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="search"
                id="search-label"
              >
                Github username
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputTypeSearch css-1g24dm6-MuiInputBase-input-MuiOutlinedInput-input"
                  id="search"
                  type="search"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1ftyaf0"
                  >
                    <span>
                      Github username
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Search
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </form>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1ri6ub7-MuiPaper-root-MuiCard-root"
        data-testid="user_card"
      >
        <img
          alt="avatar"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
          height="140"
          src="avatar.svg"
        />
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
          >
            ilyass
          </div>
        </div>
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-vkc4v8-MuiTypography-root"
          >
            followers: 
            10
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-vkc4v8-MuiTypography-root"
          >
            following: 
            20
          </p>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing makeStyles-cardActions-5 css-1t6e9jv-MuiCardActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall css-11qr2p8-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Repos
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
            href="url"
          >
            See github Profile
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
